File: algorithm_library/python/data-structure/TopK.py
class SumTopK:
  # 長さ N の配列 A の大きい方からK個の和を高速に管理する
  def __init__(self,N,A,K):
    self.A = A
    self.ok = SortedMultiset([0]*K) # 使うやつ
    self.ng = SortedMultiset([0]*(N-K)) # 使わないやつ
    self.ans = 0 # 大きい方からK個の和
  
  def add(self,y): # どこかの値がyに変わる
    self.ok.add(y)
    self.ans += y
    yy = self.ok[0]
    self.ans -= yy
    self.ok.discard(yy)
    self.ng.add(yy)
    
  def DEL(self,y): # yを削除する
    if y>=self.ok[0]:
      self.ok.discard(y)
      self.ans -= y
      yy = self.ng[-1]
      self.ng.discard(yy)
      self.ok.add(yy)
      self.ans += yy
    else:
      self.ng.discard(y)
  
  def update(self,x,y): # A[x]をyに更新し、K個の和を返す
    self.add(y)
    self.DEL(self.A[x])
    self.A[x] = y
    return self.ans

File: algorithm_library/cpp-from-py/data-structure/TopK.cpp
#include <bits/stdc++.h>
using namespace std;

struct SumTopK {
    int N, K;
    vector<long long> A;
    multiset<long long> ok, ng;
    long long ans = 0;

    SumTopK(int N_, vector<long long> A_, int K_) : N(N_), K(K_), A(move(A_)) {
        for (int i = 0; i < K; i++) ok.insert(0);
        for (int i = 0; i < N - K; i++) ng.insert(0);
    }

    void add(long long y) {
        ok.insert(y);
        ans += y;
        long long yy = *ok.begin();
        ans -= yy;
        ok.erase(ok.begin());
        ng.insert(yy);
    }

    void DEL(long long y) {
        if (K > 0 && !ok.empty() && y >= *ok.begin()) {
            auto it = ok.find(y);
            if (it != ok.end()) {
                ok.erase(it);
                ans -= y;
                if (!ng.empty()) {
                    auto jt = prev(ng.end());
                    long long yy = *jt;
                    ng.erase(jt);
                    ok.insert(yy);
                    ans += yy;
                }
            }
        } else {
            auto it = ng.find(y);
            if (it != ng.end()) ng.erase(it);
        }
    }

    long long update(int x, long long y) {
        add(y);
        DEL(A[x]);
        A[x] = y;
        return ans;
    }
};
